Mathlib/RingTheory/NoetherNormalization.lean

```diff
@@ -0,0 +1,287 @@
+private noncomputable abbrev eqv1 := quotientEquivAlg ((I.map (T f)).map
+    (finSuccEquiv k n)) (I.map (T f)) (finSuccEquiv k n).symm (by
+  set g := (finSuccEquiv k n)
+  symm
+  have : g.symm.toRingEquiv.toRingHom.comp g = RingHom.id _ :=
+    RingEquiv.symm_toRingHom_comp_toRingHom g.toRingEquiv
+  calc
+    _ = map (g.symm.toAlgHom.toRingHom.comp g) (map (T f) I) := by
+      rw [← Ideal.map_map]; rfl
+    _ = map (RingHom.id _) (map (T f) I) := by congr
+    _ = Ideal.map ((RingHom.id _).comp <| T f) I := by rw [← Ideal.map_map]; rfl
+    _ = _ := by rw [id_comp]; rfl
+  )
+
+/-$eqv2$ is the isomorphism from $k[X_0,...,X_n]/T(I)$ into $k[X_0,...,X_n]/I$,
+induce by $T$.-/
+private noncomputable abbrev eqv2 := quotientEquivAlg (R₁ := k) (I.map (T f)) I (T f).symm (by
+  symm
+  calc
+    _ = Ideal.map ((T f).symm.toRingEquiv.toRingHom.comp (T f)) I := by rw [← Ideal.map_map]; rfl
+    _ = _ := by
+      have : (T f).symm.toRingEquiv.toRingHom.comp (T f) = RingHom.id _ := by
+        apply RingEquiv.symm_toRingHom_comp_toRingHom
+      rw [this, Ideal.map_id])
+
+/-$hom2$ is the composition of maps above, from $k[X_1,...X_n]$ to $k[X_0,...X_n]/I$,
+and it's integral.-/
+private noncomputable def hom2 :=
+  (eqv2 f I).toAlgHom.comp ((eqv1 f I).toAlgHom.comp ((hom1 f I).restrictScalars k))
+
+/-Use induction to prove there is an injective map from $k[X_0,...X_{r-1}]$ to $k[X_0,...X_n]/I$.-/
+theorem exists_integral_inj (I : Ideal (MvPolynomial (Fin n) k))
+    (hi : I ≠ ⊤) : ∃ r ≤ n, ∃ g : (MvPolynomial (Fin r) k) →ₐ[k] ((MvPolynomial (Fin n) k) ⧸ I),
+    Function.Injective g ∧ g.IsIntegral := by
+  induction' n with d hd
+  · refine ⟨0, by rfl, Quotient.mkₐ k I, ?_⟩
+    constructor
+    · intro a b hab
+      rw [Quotient.mkₐ_eq_mk, Ideal.Quotient.eq] at hab
+      by_contra neq
+      apply sub_ne_zero_of_ne at neq
+      have eq := eq_C_of_isEmpty (a - b)
+      have ne : coeff 0 (a - b) ≠ 0 := fun h ↦ h ▸ eq ▸ neq <| map_zero _
+      obtain ⟨c, _, eqr⟩ := isUnit_iff_exists.mp (Ne.isUnit ne)
+      have one : c • (a - b) = 1 := by
+        rw [MvPolynomial.smul_eq_C_mul, eq, ← RingHom.map_mul, eqr, MvPolynomial.C_1]
+      exact hi ((eq_top_iff_one I).mpr (one ▸ Submodule.smul_of_tower_mem I c hab))
+    · apply isIntegral_of_surjective _ (Quotient.mkₐ_surjective k I)
+  by_cases eqi : I = 0
+  · set q := Quotient.mkₐ k I
+    have bij : Function.Bijective q := by
+      unfold q
+      simp only [Quotient.mkₐ_eq_mk]
+      constructor
+      · intro a b hab
+        rw [Ideal.Quotient.eq, eqi, Submodule.zero_eq_bot, mem_bot] at hab
+        rw [← add_zero b, ← hab, add_sub_cancel]
+      · exact Quotient.mk_surjective
+    exact ⟨d + 1, by rfl, q, bij.1, q.isIntegral_of_surjective bij.2⟩
+  · obtain ⟨f, fi, fne⟩ := Submodule.exists_mem_ne_zero_of_ne_bot eqi
+    set ϕ := kerLiftAlg <| hom2 f I
+    have ne : ker (hom2 f I) ≠ ⊤ := by
+      by_contra eq
+      replace eq := (mem_ker (f := hom2 f I) (r := 1)).mp (eq ▸ trivial)
```

Comment:
I am pretty sure we have a lemma to do this. If it is really missing we should add it.